Example/GiniVision/ComponentAPICoordinator.swift

```diff
@@ -0,0 +1,269 @@
+        if document.isReviewable {
+            showReviewScreen(withDocument: document)
+        } else {
+            showAnalysisScreen(withDocument: document)
+        }
+    }
+    
+    func didTapClose() {
+        dismissTabBarController()
+    }
+}
+
+// MARK: ComponentAPIReviewScreenDelegate
+
+extension ComponentAPICoordinator: ComponentAPIReviewScreenDelegate {
+    
+    func didReview(documentReviewed: GiniVisionDocument) {
+        // Analyze reviewed data because changes were made by the user during review.
+        if documentReviewed.data != document?.data {
+            document = documentReviewed
+            if let documentData = document?.data {
+                AnalysisManager.sharedManager.analyzeDocument(withData: documentData, cancelationToken: CancelationToken(), completion: nil)
+                showAnalysisScreen(withDocument: documentReviewed)
+            }
+            return
+        }
+        
+        // Present already existing results retrieved from the first analysis process initiated in `viewDidLoad`.
+        if let result = AnalysisManager.sharedManager.result,
+            let document = AnalysisManager.sharedManager.document {
+            handleAnalysis(result, fromDocument: document)
+            return
+        }
+        
+        // Restart analysis if it was canceled and is currently not running.
+        if !AnalysisManager.sharedManager.isAnalyzing {
+            if let documentData = document?.data {
+                AnalysisManager.sharedManager.analyzeDocument(withData: documentData, cancelationToken: CancelationToken(), completion: nil)
+            }
+        }
+        
+        // Show analysis screen if no results are in yet and no changes were made.
+        showAnalysisScreen(withDocument: documentReviewed)
+    }
+    
+    func didCancelReview() {
+        AnalysisManager.sharedManager.cancelAnalysis()
+    }
+}
+
+// MARK: ComponentAPIAnalysisScreenDelegate
+
+extension ComponentAPICoordinator: ComponentAPIAnalysisScreenDelegate {
+    func didTapErrorButton() {
+        if let document = document {
+            AnalysisManager.sharedManager.analyzeDocument(withData: document.data, cancelationToken: CancelationToken(), completion: nil)
+        }
+    }
+    
+    func didCancelAnalysis() {
+        AnalysisManager.sharedManager.cancelAnalysis()
+    }
+    
+    func didAppear() {
+        // Subscribe to analysis events which will be fired when the analysis process ends.
+        // Either with a valid result or an error.
+        NotificationCenter.default.addObserver(self, selector: #selector(handleAnalysis(errorNotification:)), name: NSNotification.Name(rawValue: GINIAnalysisManagerDidReceiveErrorNotification), object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(handleAnalysis(resultNotification:)), name: NSNotification.Name(rawValue: GINIAnalysisManagerDidReceiveResultNotification), object: nil)
```

Comment:
I hate using notifications for things like this. But I guess it was already like so. Carry on...

Reply:
Yes, I hate this too but I think it could be removed in a future.